src/query-rule-custom-data/query-rule-custom-data.ts

```diff
@@ -0,0 +1,56 @@
+import {
+  Component,
+  ContentChild,
+  TemplateRef,
+  Inject,
+  forwardRef,
+  Input,
+} from '@angular/core';
+
+import { connectQueryRules } from 'instantsearch.js/es/connectors';
+
+import { BaseWidget } from '../base-widget';
+import { NgAisInstantSearch } from '../instantsearch/instantsearch';
+
+@Component({
+  selector: 'ais-query-rule-custom-data',
+  template: `
+    <div [class]="cx()">
+      <ng-container *ngTemplateOutlet="template; context: templateContext">
+      </ng-container>
+
+      <div *ngIf="!template">
+        <div *ngFor="let item of state.items">
+          <pre>{{ item | json }}</pre>
+        </div>
+      </div>
+    </div>
+  `,
+})
+export class NgAisQueryRuleCustomData extends BaseWidget {
+  @ContentChild(TemplateRef) public template: any;
+
+  @Input() public transformItems?: <U extends {}>(items: {}[]) => U[];
```

Comment:
does actually `U` need to extend the original type for item? I'm thinking maybe having: `<U>(items: {}[]) => U[];`

Reply:
`any` is the most correct type that we can use as long as the component is not generic.

See algolia/instantsearch.js#3670.